src/java_bytecode/jar_file.cpp

```diff
@@ -1,131 +1,101 @@
 
-    for(std::size_t i=0; i<number_of_files; i++)
-    {
-      mz_uint filename_length=mz_zip_reader_get_filename(&zip, i, nullptr, 0);
-      char *filename_char=new char[filename_length+1];
-      mz_uint filename_len=
-        mz_zip_reader_get_filename(&zip, i, filename_char, filename_length);
-      assert(filename_length==filename_len);
-      std::string file_name(filename_char);
+jar_filet::jar_filet(jar_filet &&other):
+  m_zip_archive(std::move(other.m_zip_archive)),
+  m_name_to_index((other.m_name_to_index)) {}
 
-      // non-class files are loaded in any case
-      bool add_file=!has_suffix(file_name, ".class");
-      // load .class file only if they match regex / are in match set
-      add_file|=class_loader_limit.load_class_file(file_name);
-      if(add_file)
-      {
-        if(has_suffix(file_name, ".class"))
-          status() << "read class file " << file_name
-                   << " from " << filename << eom;
-        filtered_jar[file_name]=i;
-      }
-      delete[] filename_char;
-    }
-  }
+jar_filet& jar_filet::operator=(jar_filet &&other)
+{
+  m_zip_archive=std::move(other.m_zip_archive);
+  m_name_to_index=std::move(other.m_name_to_index);
+  return *this;
 }
 
-jar_filet::~jar_filet()
+std::string jar_filet::get_entry(const std::string &name)
 {
-  if(mz_ok)
+  const auto entry=m_name_to_index.find(name);
+  INVARIANT(entry!=m_name_to_index.end(), "File doesn't exist");
+  try
+  {
+    return m_zip_archive.extract(entry->second);
+  }
+  catch(const std::runtime_error &)
   {
-    mz_zip_reader_end(&zip);
-    mz_ok=false;
+    return "";
   }
 }
 
-std::string jar_filet::get_entry(const irep_idt &name)
+static bool is_space(const char ch)
 {
-  if(!mz_ok)
-    return std::string("");
-
-  std::string dest;
-
-  auto entry=filtered_jar.find(name);
-  assert(entry!=filtered_jar.end());
-
-  size_t real_index=entry->second;
-  mz_zip_archive_file_stat file_stat;
-  memset(&file_stat, 0, sizeof(file_stat));
-  mz_bool stat_ok=mz_zip_reader_file_stat(&zip, real_index, &file_stat);
-  if(stat_ok!=MZ_TRUE)
-    return std::string();
-  std::vector<char> buffer;
-  size_t bufsize=file_stat.m_uncomp_size;
-  buffer.resize(bufsize);
-  mz_bool read_ok=
-    mz_zip_reader_extract_to_mem(&zip, real_index, buffer.data(), bufsize, 0);
-  if(read_ok!=MZ_TRUE)
-    return std::string();
-
-  dest.insert(dest.end(), buffer.begin(), buffer.end());
-
-  return dest;
+  return std::isspace(ch);
 }
 
-jar_filet::manifestt jar_filet::get_manifest()
+/// Remove leading and trailing whitespace characters from string
+static std::string trim(
+  const std::string::const_iterator begin,
+  const std::string::const_iterator end)
 {
-  auto entry=filtered_jar.find("META-INF/MANIFEST.MF");
-  if(entry==filtered_jar.end())
-    return manifestt();
-
-  std::string dest=get_entry(entry->first);
-  std::istringstream in(dest);
-
-  manifestt manifest;
+  const auto out_begin=std::find_if_not(begin, end, is_space);
```

Comment:
I won't get into a tug of war with @reuk here, but just noting I think this is considerably less readable than the nice terse pair of while-loops we had before.